go/vt/vttablet/tabletmanager/vdiff/engine.go

```diff
@@ -175,7 +189,13 @@ func (vde *Engine) retry(ctx context.Context, err error) {
 // addController creates a new controller using the given vdiff record and adds it to the engine.
 // You must already have the main engine mutex (mu) locked before calling this.
 func (vde *Engine) addController(row sqltypes.RowNamedValues, options *tabletmanagerdata.VDiffOptions) error {
-	ct, err := newController(vde.ctx, row, vde.dbClientFactoryDba, vde.ts, vde, options)
+	var tmc tmclient.TabletManagerClient
+	if vde.testingTMC != nil {
```

Comment:
Why not pass `tmc` while creating the engine itself, rather than the `testingTMC` attribute?

Reply:
I removed these changes in the engine and controller and am now doing this properly in the test framework: https://github.com/vitessio/vitess/pull/11768/files#diff-eb9190e65920b52acd2700209d868f655e0308eecb2e4893ad2ff20adf74da1bR76-R77